### DISCUSSION TITLE
Load all packages in `tar_option_set()`?

### DISCUSSION THREAD

**Discussion by malcolmbarrett** [ropensci/targets#496] [Help] [answered]
Occasionally, it's useful to load all the packages that I'm using in `tar_option_set()` in my interactive session. Is there a good procedure for this? e.g. if I have `tar_option_set(packages = c("ggplot2", "gt"))`, then load ggplot2 and gt in my local interactive session

**Comment by gorkang** [comment]
Not sure what is the best way to do it, but my approach has been:

1. Run `targets::tar:renv()` once: This creates `_targets_packages()` with all the packages in `tar_option_set(packages)`
2. Using [shrtcts](https://github.com/gadenbuie/shrtcts), I have a shortcut defined (CTRL+P) that points to .shrtcts.R where I have the following:

```
message("Loading _targets_packages.R and R/helper_functions.R")

if (file.exists("./_targets_packages.R")) suppressPackageStartupMessages(source("./_targets_packages.R"))
if (file.exists("./_targets.R")) source("./_targets.R")
if (file.exists("./R/helper_functions.R")) source("./R/helper_functions.R")
```

**Reply by wlandau** [reply]
Yeah, that sounds like a great workaround, @gorkang. Alternatively, [Target Markdown](https://books.ropensci.org/targets/markdown.html) accomplishes some of this. In interactive/notebook/prototyping mode, if you run a `{targets ..., tar_globals = TRUE}` chunk that contains the `tar_option_set()` call, then any subsequent `{targets ..., tar_globals = FALSE}` chunks with actual targets will load the packages into your session automatically for interactive prototyping.

**Reply by malcolmbarrett** [reply]
Thank you both for your responses. Unfortunately, I don't think either of these solutions is what I'm looking for. I'm looking for more of a spiritual sibling to `tar_load(everything())` but to load whatever packages (or perhaps whatever prelude is in `_targets.R`) are needed. It really doesn't need to be efficient computationally the way targets normally operates with the run environment.

My perspective on this comes from beginning to teach targets in my R courses and workshops. For beginners, this is all too much overhead and too different from the mental model they have about how packages work in R. IMO, a simpler solution is really needed to help onboard people to this great tool, and I don't think that solution's design needs to be too complex. Happy to transfer this to a feature request and discuss potential solutions.

**Reply by wlandau** [reply]
This would require sourcing `_targets.R` in the interactive R session and then loading the packages. That would load the user's global objects and functions into memory, which is probably desirable from a pedagogical perspective.

```{r}
tar_load_globals <- function(
  envir = parent.frame(),
  script = targets::tar_config_get("script")
) {
  force(envir)
  source(script, local = envir)
  lapply(
    X = tar_option_get("packages"),
    FUN = library,
    character.only = TRUE,
    lib.loc = tar_option_get("library")
  )
  invisible()
}
```
 
This should be safe since `targets` defaults to a new reproducible R process for all its serious work. Happy to support this function.

**Reply by malcolmbarrett** [reply]
That is perfect! Exactly what I had in mind, and I think it is good that it sources `_targets.R` since there may be other options or files sourced there.

**Reply by wlandau** [reply]
Now implemented in dev.

**Comment by malcolmbarrett** [comment, marked as answer]
Thanks to @wlandau, the best answer to this question is now to use `tar_load_globals()`: https://github.com/ropensci/targets/commit/460e25eb0d5171f8f5d985a602c3dc6671464a55
